sdb/commands/threads.py

```diff
@@ -0,0 +1,72 @@
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import argparse
+from typing import Any, Dict, Iterable, List, Tuple
+from collections import defaultdict
+
+import drgn
+from drgn.helpers.linux.pid import for_each_task
+
+import sdb
+from sdb.commands.internal.table import Table
+from sdb.commands.stacks import Stacks
+
+class Threads(sdb.Locator, sdb.PrettyPrinter):
+    """
+    Print information about all or a single thread
+
+    task - address of the task_struct
+    state - the state of the thread
+    pid - the pid of the thread's process
+    prio - the priority of the thread
+    comm - the thread's command
+    """
+
+    names = ["threads", "thread"]
+    input_type = "struct task_struct *"
+    output_type = "struct task_struct *"
+
+    @classmethod
+    def _init_parser(cls, name: str) -> argparse.ArgumentParser:
+        parser = super()._init_parser(name)
+        return parser
+
+    FIELDS = {
+        "task": lambda obj: hex(obj.value_()),
+        "state": lambda obj: str(Stacks.task_struct_get_state(obj)),
+        "pid": lambda obj: int(obj.pid),
+        "prio": lambda obj: int(obj.prio),
+        "comm": lambda obj: str(obj.comm.string_().decode("utf-8")),
+    }
+
+    def pretty_print(self, objs: Iterable[drgn.Object]) -> None:
+        fields = list(Threads.FIELDS.keys())
+        table = Table(fields, None, {"task" : str})
+        for obj in objs:
+            row_dict = {
+                field: Threads.FIELDS[field](obj) for field in fields
+            }
+            table.add_row(row_dict["task"], row_dict)
+        table.print_()
+
+    def no_input(self) -> Iterable[drgn.Object]:
+        task_list = list(for_each_task(sdb.get_prog()))
```

Comment:
Rather than constructing the list, does `for_each_task()` return an iterable suitable for passing to execute_pipeline()?  I think that would avoid staging all the tasks here; instead we would gather the tasks as we go.